sql: introduce proto with metadata needed by fetchers, use it in row.Fetcher

RaduBerinde opened the pull request:
#### descpb: introduce IndexFetchSpec

This commit introduces a new IndexFetchSpec proto which efficiently
encodes the necessary metadata to fetch a set of columns from an
index. The goal is to use this instead of serializing the entire
TableDescriptor when we execute a distributed scan.

Release note: None

#### sql: use IndexFetchSpec in row.Fetcher

In this commit, most of the row.Fetcher code is modified to internally
use an IndexFetchSpec instead of the table and index descriptor.

The remaining use is around decoding keys when converting errors.

One interesting thing to note here is that the row fetcher no longer
relies on the caller to "fix up" inverted column types; we do this
automatically now when creating the `IndexFetchSpec`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75633)
<!-- Reviewable:end -->

ajwerner commented on pkg/sql/rowenc/index_fetch.go:
```diff
@@ -0,0 +1,146 @@
+// inverted and we fetch the inverted key, the corresponding Column contains the
+// inverted column type.
+func InitIndexFetchSpec(
+	s *descpb.IndexFetchSpec,
```
I get why you're doing this, but would it be better to make this a method on catalog.TableDescriptor and cache them on the leased descriptor wrapper objects? Seems like this thing is immutable and avoiding this re-computation and allocations seems good. As of https://github.com/cockroachdb/cockroach/pull/74722 we generally get a handle to the leased descriptor

ajwerner reviewed (commented):
> The goal is to use this instead of serializing the entire
TableDescriptor when we execute a distributed scan.

Wouldn't it be even better to just serialize a descriptor ID, version, timestamp triple? https://github.com/cockroachdb/cockroach/issues/74352 That doesn't feel very far off. 

Do you have bigger dreams for this thing? If it's just about optimizing distsql messages, I'm far from sold. I've heard you suggest crazier push down ideas. Is this part of that?

RaduBerinde reviewed (commented):
The bigger dream is to revive Jordan's prototype where KV builds the columnar batches (#52863). There we are talking about many more,  smaller KV operations (max ~10MB of data scanned at a time); serializing and deserializing the table descriptor would be a much bigger problem than in DistSQL today (where it essentially happens once per query per node). As for having the descriptor available by ID/version on the KV side, that seems very problematic in multi-tenant (as well as architecturally).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-:-MuSqx_6AEQmnvK3cseo:b-7ypfpa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @samiskin, and @yuzefovich)

___
*[pkg/sql/rowenc/index_fetch.go, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MuSfd3LiEYXmoXhWhGv-r2-31:-MuSt1bp58s_in5iJDux:b76vvg0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3acef1b0ea1c2988059f19cccd11b6a2d876886a/pkg/sql/rowenc/index_fetch.go#L31)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I get why you're doing this, but would it be better to make this a method on catalog.TableDescriptor and cache them on the leased descriptor wrapper objects? Seems like this thing is immutable and avoiding this re-computation and allocations seems good. As of https://github.com/cockroachdb/cockroach/pull/74722 we generally get a handle to the leased descriptor
</blockquote></details>

Most of the spec (other than the fetched columns) is indeed immutable and we could have it cached in the descriptor. On the other hand, we're only talking about populating some fields and allocating one slice (`KeyColumns`) so it's not very pressing.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> As for having the descriptor available by ID/version on the KV side, that seems very problematic in multi-tenant (as well as architecturally).

We're aligned that the ID/Version in the spec thing is just for sql pods. I just wanted to confirm that you're not doing this work to diminish the number of allocations on the sql-pod side in distsql (as is mentioned in the commit messages). 

> There we are talking about many more, smaller KV operations (max ~10MB of data scanned at a time); serializing and deserializing the table descriptor would be a much bigger problem than in DistSQL today (where it essentially happens once per query per node).

No doubt. I think what I'd do is:
1) box this widdled down spec in bytes in roachpb protos to sidestep import sadness 
2) cache the fully serialized spec (and deserialized) spec in the `catalog.TableDescriptor` object we cache in individual sql pods (as suggested). 
3) on the kvserver-side inject interface that takes those bytes and gives you back some column batch builder thingy that you can pass to MVCC. That interface internally can have a cache of serialized spec to deserialized things. 
4) (optional) do the (ID/Version) thing in distsql specs 

That way, in the common case, nobody is ever serializing or deserializing a spec or table descriptor ~anywhere. Just shuttling some bytes as a cache key. And, as you noted, your widdled down `IndexFetchSpec` is going to be pretty small. 

Seems like all of this can be done without too too much in the way of hideous boundary violation or cost. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-:-MuSvD3mCW2ZP8yxJ9c0:b-ycrur9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @samiskin, and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Nice! I only have nits, but it'd probably be good to get an approval from someone from Schema too. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 7 files at r1, 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-:-MulMD6p-y-jQ676tKZc:blu47uy)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @RaduBerinde, and @samiskin)

___
*[pkg/sql/catalog/descpb/index_fetch.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulFiU-E29SeK_TPa_U:-MulFiU04Ic4p-ljdFRw:b-4igc5m) ([raw file](https://github.com/cockroachdb/cockroach/blob/058e0140f6dc6977be96c7f635383bbd423cf237/pkg/sql/catalog/descpb/index_fetch.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

nit: `s/2017/2022/`.
___
*[pkg/sql/catalog/descpb/index_fetch.proto, line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulGCHS-beUzYzji298:-MulGCHS-beUzYzji299:b-vjtmq8) ([raw file](https://github.com/cockroachdb/cockroach/blob/058e0140f6dc6977be96c7f635383bbd423cf237/pkg/sql/catalog/descpb/index_fetch.proto#L26)):*
> ```protobuf
>   message Column {
>     optional string name = 1 [(gogoproto.nullable) = false];
>     optional uint32 column_id = 2 [(gogoproto.nullable) = false, (gogoproto.customname) = "ColumnID", (gogoproto.casttype) = "ColumnID"];
> ```

nit: long line.
___
*[pkg/sql/catalog/descpb/index_fetch.proto, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulG2asE2EMD7vHU-JN:-MulG2asE2EMD7vHU-JO:b8tyikc) ([raw file](https://github.com/cockroachdb/cockroach/blob/058e0140f6dc6977be96c7f635383bbd423cf237/pkg/sql/catalog/descpb/index_fetch.proto#L33)):*
> ```protobuf
>     optional sql.sem.types.T type = 3;
> 
>     // IsNonNullable indicates that it would be an corruption error if we ever
> ```

nit: `s/an/a/`.
___
*[pkg/sql/catalog/descpb/index_fetch.proto, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulGTEK2_Bjpa7MgiFR:-MulGTEK2_Bjpa7MgiFS:blgvpya) ([raw file](https://github.com/cockroachdb/cockroach/blob/058e0140f6dc6977be96c7f635383bbd423cf237/pkg/sql/catalog/descpb/index_fetch.proto#L78)):*
> ```protobuf
>                                      (gogoproto.casttype) = "IndexDescriptorEncodingType"];
> 
>   // NumKeySuffixColumns returns the number of suffix columns (corresponding to
> ```

nit: "returns" sounds odd.
___
*[pkg/sql/row/fetcher.go, line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulJHmx0Vjtq7TcbZpO:-MulJHmx0Vjtq7TcbZpP:bd3hni9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2317a7bd6b06fbc613c40b953c8a15bfc85c73d/pkg/sql/row/fetcher.go#L82)):*
> ```Go
> 
> 	// One value per column that is part of the key; each value is a column
> 	// index (into FetchedColumns); -1 if we don't need the value for that column.
> ```

nit: "cols" is also mentioned in the comments on `neededValueColsByIdx` and `colIdxMap`.
___
*[pkg/sql/row/fetcher.go, line 868 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulKfEt5ugUu6LzRUBR:-MulKfEt5ugUu6LzRUBS:b-h96ga6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73be01b43b977e732bbe8f3bfc460c218a0799/pkg/sql/row/fetcher.go#L868)):*
> ```Go
> 		}
> 
> 		if DebugRowFetch {
> ```

nit: do we not want to keep it?
___
*[pkg/sql/row/fetcher.go, line 1122 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulLVc-7UkXtULcaAuC:-MulLVc-7UkXtULcaAuD:bh8xnc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2317a7bd6b06fbc613c40b953c8a15bfc85c73d/pkg/sql/row/fetcher.go#L1122)):*
> ```Go
> 				for _, idx := range table.indexColIdx {
> 					if idx != -1 {
> 						indexColValues = append(indexColValues, table.row[idx].String(table.spec.FetchedColumns[idx].Type))
> ```

nit: we probably can use `col.Type` here.
___
*[pkg/sql/rowenc/index_encoding.go, line 511 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulLuYr8qvvLZPQVSSV:-MulLuYr8qvvLZPQVSSW:b8vnrhp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2317a7bd6b06fbc613c40b953c8a15bfc85c73d/pkg/sql/rowenc/index_encoding.go#L511)):*
> ```Go
> 			return nil, false, err
> 		}
> 		if vals[j].IsNull() {
> ```

nit: it should be slightly faster with
```
foundNull = foundNull || vals[j].IsNull()
```
___
*[pkg/sql/rowenc/testdata/index-fetch, line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulHzzC4f2_SpluxsWS:-MulHzzC4f2_SpluxsWT:b-62ayx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/058e0140f6dc6977be96c7f635383bbd423cf237/pkg/sql/rowenc/testdata/index-fetch#L71)):*
> ```
> # Primary index scan, not all columns.
> index-fetch
> table: t 
> ```

nit: trailing spaces.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
descpb: introduce IndexFetchSpec

This commit introduces a new IndexFetchSpec proto which efficiently
encodes the necessary metadata to fetch a set of columns from an
index. The goal is to use this instead of serializing the entire
TableDescriptor when we execute a distributed scan.

Release note: None

RaduBerinde committed:
sql: use IndexFetchSpec in row.Fetcher

In this commit, most of the row.Fetcher code is modified to internally
use an IndexFetchSpec instead of the table and index descriptor.

The remaining use is around decoding keys when converting errors.

One interesting thing to note here is that the row fetcher no longer
relies on the caller to "fix up" inverted column types; we do this
automatically now when creating the `IndexFetchSpec`.

Release note: None

RaduBerinde reviewed (commented):
TFTR!

> it'd probably be good to get an approval from someone from Schema too

@ajwerner sounds like you are ok with this?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-:-MulVO_b1Eyd3n_ludpZ:bos36ga)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @samiskin, and @yuzefovich)

___
*[pkg/sql/row/fetcher.go, line 868 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulKfEt5ugUu6LzRUBR:-MulTnJB6eAM9v-5zZbk:bxsfj0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73be01b43b977e732bbe8f3bfc460c218a0799/pkg/sql/row/fetcher.go#L868)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: do we not want to keep it?
</blockquote></details>

I'm not convinced there's much value to these (KV traces have more info).
___
*[pkg/sql/row/fetcher.go, line 1122 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulLVc-7UkXtULcaAuC:-MulTPgG-HCpvnL148mj:bl5ud8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2317a7bd6b06fbc613c40b953c8a15bfc85c73d/pkg/sql/row/fetcher.go#L1122)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we probably can use `col.Type` here.
</blockquote></details>

It's a different index


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 3 of 12 files at r3, 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75633)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @RaduBerinde, and @samiskin)

___
*[pkg/sql/row/fetcher.go, line 868 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-MulKfEt5ugUu6LzRUBR:-MuldSWl5EGsxIMhn5vQ:blagbmq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73be01b43b977e732bbe8f3bfc460c218a0799/pkg/sql/row/fetcher.go#L868)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I'm not convinced there's much value to these (KV traces have more info).
</blockquote></details>

I think I agree - let's just remove it altogether then.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> @ajwerner sounds like you are ok with this?

Indeed

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75633#-:-MulfRXz9lM5bwYC1Vux:b9zi69n)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @RaduBerinde, and @samiskin)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
CI failure is a test flake.

bors r+
